### PR TITLE
Allow improvements that don't need removal to build

### LINKED ISSUE #11296: "Does not need removal of [tileFilter]" not accepting "All" or "Terrain" as filters
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Game Version

4.10.17

### Describe the bug

I attempted to make a tile improvement that's buildable anywhere, including on top of jungles and forests and marshes. I did this by giving it the unique of `"Does not need removal of [All]"`, hoping that this would cover all possible cases, instead of having to make a unique for every single kind of terrain. It is not working. This unique fails to make the improvement buildable on unbuildable terrain features.

### Steps to Reproduce

I've also tested this with `"Does not need removal of [Terrain]"` and `"Does not need removal of [Land]"`. None of these produce the desired effect.

### Screenshots

_No response_

### Link to save file

This mod contains a simple "Test" tile improvement that's buildable on Land and includes the unique discussed above.
[ImprovementTest.zip](https://github.com/yairm210/Unciv/files/14563852/ImprovementTest.zip)

### Operating System

Windows

### Additional Information

This issue also shows up in the Bio-Tech-aliens mod by @Doomstar00.

### PR DESCRIPTION
**Pull request by SeventhM**
closes #11296

Also allows for Tile Feature removal to refer to features by filter. This opens a bit of a can of worms as we now have more or less 3 tile filter options:
1. This new version of TerrainFilter (filters for stuff inherent to the terrain)
2. The old TerrainFilter (filter for terrain on a tile
3. TileFilter (filters for any additional tile information, namely improvements)

There's also a secondary change thrown in since I noticed it (kinda as a red herring tbh) when I was adding in the TerrainFilter. Apparently, improvements that automatically removed removable terrain only checked if removing the top layer was valid, rather then all layers, which is extremely inconsistent if a mod wanted to abuse that fact. Main PR is rather small, and this fix is likely unknown (how many mods have stacking removable features?) so I have no issues myself sneaking it in

### CODE REVIEW

**Comment by yairm210** [comment]
Regarding general terrain filter - 👍🏿 
Regarding removable feature decision tree - let's add some tests for this

**Comment by SeventhM** [comment]
Sorry, was busy working on something else. Though... Seems like many of the tests here aren't particularly fully modern and/or they're more testing the base ruleset rather than that it claims to test

I'm still not sure if/how we should update what we claim "terrainFilter" is in our documentation/unique placeholder text given now there's the three distinctions. TileFilter (current definition) and TerrainFilter (this new definition) seems obvious, but something like a "TileTerrainFilter" seems like it might still be necessary. I could be wrong, though

File: tests/src/com/unciv/logic/map/TileImprovementConstructionTests.kt [hunk 1]
```diff
@@ -214,6 +213,38 @@ class TileImprovementConstructionTests {
...
+        Assert.assertFalse("Forest doesn't allow building unless allowed",
+            tile.improvementFunctions.canBuildImprovement(improvement, civInfo))
+
+
+        val allowedImprovement = testGame.createTileImprovement()
+        allowedImprovement.terrainsCanBeBuiltOn += "Forest"
+        Assert.assertTrue("Forest should allow building when allowed",
+            tile.improvementFunctions.canBuildImprovement(improvement, civInfo))
+        tile.changeImprovement(allowedImprovement.name)
+        Assert.assertTrue(tile.improvement == allowedImprovement.name)
```
Comment (yairm210) [inline tests/src/com/unciv/logic/map/TileImprovementConstructionTests.kt:232]: Assert forest exists?

**Review by yairm210 — APPROVED** [review]

File: core/src/com/unciv/models/ruleset/tile/Terrain.kt [hunk 2]
```diff
@@ -143,6 +144,29 @@ class Terrain : RulesetStatsObject() {
...
+            name -> true
+            "Terrain" -> true
+            in Constants.all -> true
+            "Open terrain" -> !isRough()
+            "Rough terrain" -> isRough()
+            type.name -> true
+            "Natural Wonder" -> type == TerrainType.NaturalWonder
+            "Terrain Feature" -> type == TerrainType.TerrainFeature
+
+            else -> {
```
[the remark this answers is not in the document]
Comment (SeventhM) [reply core/src/com/unciv/models/ruleset/tile/Terrain.kt]: Copy pasted from TileFilter (and deleted what wasn't relevant). Though, In the future, maybe in the future we might end up needing the brackets

**Pull request merged**